contrib/raftexample/raftchan.go

```diff
@@ -0,0 +1,240 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"net/http"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/etcdserver/stats"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/coreos/etcd/rafthttp"
+	"github.com/coreos/etcd/wal"
+	"github.com/coreos/etcd/wal/walpb"
+)
+
+type raftKV struct {
+	Key	string
+	Val	[]byte
+}
+
+// A key-value stream backed by raft
+type raftKVChan struct {
+	proposeC	<-chan raftKV	// proposed messages (k,v)
+	commitC		chan raftKV	// entries committed to log (k,v)
+	errorC		chan error	// errors from raft session
+
+	id		int		// client ID for raft session
+	peers		[]string	// raft peer addresses
+
+	// raft backing for the commit/error channel
+	node		raft.Node
+	raftStorage	*raft.MemoryStorage
+	wal		*wal.WAL
+	transport	*rafthttp.Transport
+}
+
+// A raft channel represents a key-value raft workflow over http using channels:
+// * client sends (k,v) change proposals by writing to a proposal channel.
+// * client receives a stream of commits by reading a commit channel.
+//   On initialization, the write ahead log replays into the commit channel; to
+//   avoid exposing stale values, a nil value over the commit channel signals
+//   the replay is complete.
+// * client receives raft errors via an out-of-band error channel.
+func newRaftChan(id int, peers []string, proposeC <-chan raftKV) (<-chan raftKV, <-chan error) {
+	rc := &raftKVChan {
+		id:		id,
+		peers:		peers,
+		proposeC:	proposeC,
+		commitC:	make(chan raftKV),
+		errorC:		make(chan error),
+		raftStorage:	raft.NewMemoryStorage(),
+		// rest of structure populated after WAL replay
+	}
+	go rc.startRaft()
+	return rc.commitC, rc.errorC
+}
+
+// write request entries to commit channel
+func (rc *raftKVChan) publishEntries(entries []raftpb.Entry) {
+	for i := range entries {
+		req := etcdserverpb.Request{}
+		err := req.Unmarshal(entries[i].Data)
+		if err == nil && req.Method == "PUT" {
+			// only process request entries
+			rc.commitC <- raftKV{Key: req.Path, Val: []byte(req.Val)}
+		}
+	}
+}
+
+// Open/Create WAL for replay/reading
+func (rc *raftKVChan) openWAL() *wal.WAL {
+	waldir := fmt.Sprintf("raftexample-%d", rc.id)
+	if wal.Exist(waldir) == false {
+		if err := os.Mkdir(waldir, 0750); err != nil {
+			log.Fatalf("raftexample: cannot create dir for wal (%v)", err)
+		}
+
+		w, err := wal.Create(waldir, nil)
+		if err != nil {
+			log.Fatalf("raftexample: create wal error: %v", err)
```

Comment:
let's keep the error message consistent with the previous one. `raftexample: create wal error (%v)` instead of `raftexample: create wal error :%v`